pkg/event/config.go

```diff
@@ -0,0 +1,292 @@
+type Queue queue
+
+// UnmarshalXML - decodes XML data.
+func (q *Queue) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
+	parsedQueue := queue{}
+	if err := d.DecodeElement(&parsedQueue, &start); err != nil {
+		return err
+	}
+
+	if len(parsedQueue.Events) == 0 {
+		return errors.New("missing event name(s)")
+	}
+
+	eventStringSet := set.NewStringSet()
+	for _, eventName := range parsedQueue.Events {
+		if eventStringSet.Contains(eventName.String()) {
+			return &ErrDuplicateEventName{eventName}
+		}
+
+		eventStringSet.Add(eventName.String())
+	}
+
+	*q = Queue(parsedQueue)
+
+	return nil
+}
+
+// Validate - checks whether queue has valid values or not.
+func (q Queue) Validate(region string, targetList *TargetList) error {
+	if region != "" && q.ARN.region != region {
+		return &ErrUnknownRegion{q.ARN.region}
+	}
+
+	if !targetList.Exists(q.ARN.TargetID) {
+		return &ErrARNNotFound{q.ARN}
+	}
+
+	return nil
+}
+
+// SetRegion - sets region value to queue's ARN.
+func (q *Queue) SetRegion(region string) {
+	q.ARN.region = region
+}
+
+// ToRulesMap - converts Queue to RulesMap
+func (q Queue) ToRulesMap() RulesMap {
+	pattern := q.Filter.RuleList.Pattern()
+	return NewRulesMap(q.Events, pattern, q.ARN.TargetID)
+}
+
+// Unused.  Available for completion.
+type lambda struct {
+	common
+	ARN string `xml:"CloudFunction"`
+}
+
+// Unused. Available for completion.
+type topic struct {
+	common
+	ARN string `xml:"Topic" json:"Topic"`
+}
+
+type config struct {
+	XMLName    xml.Name `xml:"NotificationConfiguration"`
+	QueueList  []Queue  `xml:"QueueConfiguration,omitempty"`
+	LambdaList []lambda `xml:"CloudFunctionConfiguration,omitempty"`
+	TopicList  []topic  `xml:"TopicConfiguration,omitempty"`
+}
+
+// Config - notification configuration described in
+// http://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html
+type Config config
+
+// UnmarshalXML - decodes XML data.
+func (conf *Config) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
+	parsedConfig := config{}
+	if err := d.DecodeElement(&parsedConfig, &start); err != nil {
+		return err
+	}
+
+	if len(parsedConfig.QueueList) == 0 {
+		return errors.New("missing queue configuration(s)")
+	}
+
+	for i, q1 := range parsedConfig.QueueList[:len(parsedConfig.QueueList)-1] {
+		for _, q2 := range parsedConfig.QueueList[i+1:] {
+			if reflect.DeepEqual(q1, q2) {
+				return &ErrDuplicateQueueConfiguration{q1}
+			}
+		}
+	}
+
+	if len(parsedConfig.LambdaList) > 0 || len(parsedConfig.TopicList) > 0 {
+		return &ErrUnsupportedConfiguration{}
+	}
+
+	*conf = Config(parsedConfig)
+
+	return nil
+}
+
+// Validate - checks whether config has valid values or not.
+func (conf Config) Validate(region string, targetList *TargetList) error {
+	for _, queue := range conf.QueueList {
+		if err := queue.Validate(region, targetList); err != nil {
+			return err
+		}
+
+		// TODO: Need to discuss/check why same ARN cannot be used in another queue configuration.
```

Comment:
There is nothing in the docs that say we can't have same queue ARN shared between queue configurations as long. It is invalid (only) to have overlapping prefixes for the same event. IIUC, configurations should have unique 'actions' for every event, prefix, suffix combination. i.e, for a given event and prefix (and/or suffix) there should be only one rule that applies to it.

Reply:
Current code does this check and wanted to know why it is a problem.  As S3 documentation says nothing about this case, I don't know whether it is a failure case in S3.  Logically I see no issues supporting it.